### PR TITLE
Label to VHost and Dynamic environments migration

### PR DESCRIPTION
**Pull request by renuka-fernando**
## Purpose
$subject

## Approach
- Read labels and label's access URLs
- Convert label to a dynamic environment
- Convert access URL to a VHost
- Delete Labels and Label URLs. (these tables should be dropped, will carry on this by fixing server usage of these tables)
- Deploy API revisions to a dynamic environment with the first VHost of the environment if the API is assigned to a label

## Related Issues
https://github.com/wso2/product-apim/issues/9807

### CODE REVIEW

**Review by CrowleyRajapakse — APPROVED** [review]

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java [hunk 1]
```diff
@@ -826,4 +844,123 @@ public static String getStringFromInputStream(InputStream is) throws APIMigratio
...
+                    int id = -1;
+                    if (rs.next()) {
+                        id = rs.getInt(1);
+                    }
+                    addGatewayVhosts(conn, id, environment.getVhosts());
+                }
+                conn.commit();
+            } catch (SQLException ex) {
+                conn.rollback();
+                throw new APIMigrationException("Failed to add data to AM_API_RESOURCE_SCOPE_MAPPING table : ", ex);
```
Comment (isharac) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java:909]: error message is misleading

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java [hunk 2]
```diff
@@ -826,4 +844,123 @@ public static String getStringFromInputStream(InputStream is) throws APIMigratio
...
+            }
+        } catch (SQLException ex) {
+            throw new APIMigrationException("Failed to add data to AM_API_RESOURCE_SCOPE_MAPPING table : ", ex);
```
Comment (isharac) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java:912]: here as well

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java [hunk 3]
```diff
@@ -826,4 +844,123 @@ public static String getStringFromInputStream(InputStream is) throws APIMigratio
...
+                prepStmt.setString(3, vhost.getHttpContext());
+                prepStmt.setString(4, vhost.getHttpPort().toString());
+                prepStmt.setString(5, vhost.getHttpsPort().toString());
+                prepStmt.setString(6, vhost.getWsPort().toString());
+                prepStmt.setString(7, vhost.getWssPort().toString());
+                prepStmt.addBatch();
+            }
+            prepStmt.executeBatch();
+        } catch (SQLException e) {
+            throw new APIMigrationException("Failed to add data to AM_API_RESOURCE_SCOPE_MAPPING table : ", e);
```
Comment (isharac) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java:939]: check for the error message

**Review by isharac — CHANGES_REQUESTED** [review]

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java (hunk 1 above)
Comment (renuka-fernando) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java]: fixed with https://github.com/wso2-extensions/apim-migration-resources/pull/85/commits/72452355b510866a43d5c954191389ad38475a2b

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java (hunk 3 above)
Comment (renuka-fernando) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java]: fixed with https://github.com/wso2-extensions/apim-migration-resources/pull/85/commits/72452355b510866a43d5c954191389ad38475a2b

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java (hunk 2 above)
Comment (renuka-fernando) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dao/APIMgtDAO.java]: fixed with https://github.com/wso2-extensions/apim-migration-resources/pull/85/commits/72452355b510866a43d5c954191389ad38475a2b

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dto/GatewayEnvironmentDTO.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
+package org.wso2.carbon.apimgt.migration.dto;
```
Comment (pubudu538) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dto/GatewayEnvironmentDTO.java:1]: Let's add missing license headers.

Comment (renuka-fernando) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/dto/GatewayEnvironmentDTO.java:19]: fixed with https://github.com/wso2-extensions/apim-migration-resources/pull/85/commits/a907847e46a736efceaeac98407c2c9e904e4784

**Review by pubudu538 — APPROVED** [review]

**Review by isharac — APPROVED** [review]

**Pull request merged**